charmcraft/providers.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Build environment provider support for charmcraft."""
+
+import logging
+import subprocess
+from typing import Optional
+
+from craft_providers import Executor, bases
+from craft_providers.actions import snap_installer
+
+logger = logging.getLogger(__name__)
+
+
+class CharmcraftBuilddBaseConfiguration(bases.BuilddBase):
+    """Base configuration for Charmcraft.
+
+    :cvar compatibility_tag: Tag/Version for variant of build configuration and
+        setup.  Any change to this version would indicate that prior [versioned]
+        instances are incompatible and must be cleaned.  As such, any new value
+        should be unique to old values (e.g. incrementing).
+    :cvar instance_config_path: Path to persistent environment configuration
+        used for compatibility checks (or other data).  Set to
+        /etc/craft-instance.conf, but may be overridden for application-specific
+        reasons.
+    :cvar instance_config_class: Class defining instance configuration.  May be
+        overridden with an application-specific subclass of InstanceConfiguration
+        to enable application-specific extensions.
+
+    :param alias: Base alias / version.
+    :param environment: Environment to set in /etc/environment.
+    :param hostname: Hostname to configure.
+    """
```

Comment:
I find weird to have all this documentation about variables and __init__ parameters when they are not really used here.

I understand that because of inheriting `bases.BuilddBase`, the _init__ params are the same... but the doc is super decoupled from the code! If the code in `bases.BuilddBase.__init__` is changed in the future nobody will remember to update the code here, for example...

Reply:
Done! Updated the docs to remove things not specific to Charmcraft.